### PR TITLE
[G2M] dropdown - add auto complete variant & onClick select styling

### PR DESCRIPTION
**Pull request by kc-leung**
- Added dropdown menu autoComplete textfield variant
- Added item active state on pointer down styling

### CODE REVIEW

File: src/base-components/dropdown-base/index.js [hunk 1]
```diff
@@ -102,11 +103,13 @@ const DropdownBase = forwardRef(({
   return (
     <div ref={ref} className={containerClasses.root} {...rest}>
       <Menu>
-        <Menu.Button as="div">
-          <div className={`${button ? 'button-container' : dropdownClasses.button}`} onClick={onClick}>
-            {button ? button : _button}
-          </div>
-        </Menu.Button>
+        {customTrigger ? customTrigger : 
+          <Menu.Button as="div">
```
Comment (woozyking) [inline src/base-components/dropdown-base/index.js:107]: would an `or` logic work here?

```jsx
{customTrigger || (
  <Menu.Button as="div">
  ...
)}
```

File: src/base-components/input-base.js [hunk 2]
```diff
@@ -69,8 +70,8 @@ const InputBase = forwardRef(({
         {...rest}
       />
       {suffix && <span className={classes.suffix}>{suffix}</span>}
-      {endIcon && !_value && !deleteButton && <div className={classes.endIcon}>{endIcon}</div>}
-      {_value && deleteButton && 
+      {endIcon && (value ? !value : !_value) && <div className={classes.endIcon}>{endIcon}</div>}
+      {deleteButton && (value ? value : _value) &&
```
Comment (woozyking) [inline src/base-components/input-base.js:74]: similarly, `value || _value`

File: src/components/dropdown-auto-complete.js [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    setFilteredOptions([])
+  }
+
+  const onChange = (val) => {
+    const filteredSuggestions = options.filter(
+      suggestion =>
+        suggestion.title.toLowerCase().indexOf(val.toLowerCase()) > -1,
+    )
+    setFilteredOptions(filteredSuggestions)
+    !val && setSelectedOptions([])
```
Comment (woozyking) [inline src/components/dropdown-auto-complete.js:169]: if the evaluation result is not assigned in a variable, prefer to use explicit flow control:

```js
if (!val) {
  setSelectedOptions([])
}
```

**Review by woozyking — COMMENTED** [review]
one usage issue would be that no-match is currently a match-all. not sure if this is intended by design and would like to understand the rationale if so

no result

**Comment by sallkall** [comment]
> one usage issue would be that no-match is currently a match-all. not sure if this is intended by design and would like to understand the rationale if so
> 
> no result

No not by design, but I never made an example for this edge case - my bad. @kc-leung should we just hide/close the dropdown if there is no match? The user can delete characters and when there is a match it can reappear

**Review by woozyking — APPROVED** [review]

**Comment by woozyking** [comment]
https://user-images.githubusercontent.com/2837532/132735052-a5eb79de-279d-4875-9f5e-135a5f947dcd.mov

2 things from ^ recording:
1. Can't figure out how exactly to reproduce the behavior where auto-complete fail to function, but it seemed to correlate with an action of taking screenshots (either Firefox `cmd + shift + S` or Mac OS native `cmd + shift + 5`) -- not sure if it's just coincidence or there's actual causation
2. Minor issue on search not reflecting the categories (not sure if intended?)

another issue is when a item is selected, the categories remain in the options (but can't be selected, naturally)
cats

**Comment by kc-leung** [comment]
> 2021-09-09.13.34.40.mov
> 
> 2 things from ^ recording:
> 
>     1. Can't figure out how exactly to reproduce the behavior where auto-complete fail to function, but it seemed to correlate with an action of taking screenshots (either Firefox `cmd + shift + S` or Mac OS native `cmd + shift + 5`) -- not sure if it's just coincidence or there's actual causation
> 
>     2. Minor issue on search not reflecting the categories (not sure if intended?)
> 
> 
> another issue is when a item is selected, the categories remain in the options (but can't be selected, naturally)
> cats

category is not a selectable item which is why it is not searchable. Also, it's in an upper level of the array, it defines the key for the item. If we want it to be searchable I might have to change the data structure for the dropdown menu. I can't replicate the behaviour from the video :(

**Review by sallkall — APPROVED** [review]

**Comment by woozyking** [comment]
> > another issue is when a item is selected, the categories remain in the options (but can't be selected, naturally)
> > cats

can we hide the categories then if there aren't any matches under it?

> category is not a selectable item which is why it is not searchable. Also, it's in an upper level of the array, it defines the key for the item. If we want it to be searchable I may have to change the data structure for the dropdown menu. I can't replicate the behaviour from the video :(

changing the data structure may not be a bad idea but we can defer to the next iteration

**Comment by kc-leung** [comment]
> > > another issue is when a item is selected, the categories remain in the options (but can't be selected, naturally)
> > > cats
> 
> can we hide the categories then if there aren't any matches under it?
> 
> > category is not a selectable item which is why it is not searchable. Also, it's in an upper level of the array, it defines the key for the item. If we want it to be searchable I may have to change the data structure for the dropdown menu. I can't replicate the behaviour from the video :(
> 
> changing the data structure may not be a bad idea but we can defer to the next iteration

Changed the data structure internally within the component using a separate array so there's no need to change it's original data structure. Now there is a working search for category and also hiding them when it is empty

**Review by woozyking — APPROVED** [review]

**Pull request merged**